flips/20220203-capability-controllers.md

```diff
@@ -0,0 +1,458 @@
+
+In the private example the call would change to `issuer.unlink<&{HasCount}>(/private/hasCount)`. It is important to note that if the issuer wants to have the ability to revoke/redirect capabilities in a more granular way (instead of doing them all at once), the solution is to create multiple private linked paths (e.g. `/private/hasCountAlice`, `/private/hasCountBob`).
+
+If a path that was unlinked is linked again all the capabilities created from that path resume working. This can be used to redirect a capability to a different object, but can also be dangerous if done unintentionally, reviving a previously revoked capability.
+
+### Capabilities are values
+
+Capabilities are a value type. This means that any capability that an account has access to can be copied, and potentially given to someone else. The copied capability will use the link on the same path as the original capability
+
+## Problem statement
+
+The current capability API has a few pain points: 
+
+- The concepts capabilities/linking/revoking/redirecting are hard to understand for new developers that are coming to Flow, as that is a lot of new concepts to grasp before one can get started.
+
+- Issuing and managing private capabilities that can be revoked at a granular level, by creating a custom private linked path for each capability, is difficult.
+
+- Unless the path name includes some hints, there is no way to know when a path was created, thus making it more difficult to remember why a certain capability was issued.
+
+- If an old path (that has been unlinked) is accidentally reused and re-linked, this will revive a capability that is supposed to be revoked.
+
+## Suggested change
+
+The suggested change addresses these pain points by:
+
+- Removing (or abstracting away) the concept of links.
+- Making it easier to create new capabilities (with individual links).
+- Offering a way to iterate through capabilities on a storage path.
+- Removing the need to have a `/private` domain.
+- Introducing Capability Controllers (CapCons) that handle management of capabilities.
+
+### Capability Controllers (CapCons)
+
+Each Capability would have an associated CapCon that would be responsible for managing the Capability. The CapCon would be created when the Capability is issued (created). If the Capability is copied (since it is a value type) it shares the CapCon with the original (i.e.: calling revoke on the CapCon revokes all copies of the associated capability).
+
+The data associated with CapCons would be stored in arrays, so that each storage path on an account has an array of CapCons of Capabilities issued from that storage path. CapCons are a non-storable object (similar to AuthAccounts).
+
+The definition of the `CapabilityController`.
+
+```cadence
+struct CapabilityController {
+   // The block height when the capability was created.
+   let issueHeight: UInt64
+   // The Type of the capability.
+   let capabilityType: Type
```

Comment:
Maybe we can make the name more descriptive? The type indicates how the capability can be borrowed – maybe `borrowType`? This is what we already use internally in the Cadence implementation.

Also, maybe this field should be mutable? One problem that came up before that changing how something in storage can be borrowed requires removing the link and re-linking.

Reply:
borrowType sounds better. I will edit that.

Not sure about mutability. Changing the borrowType on the CapCon would also mean that the existing capabilities now use a different borrow type... not sure if that is ok. I also think this might be an addition that can be made later.

Reply:
Good point, agreed!